### PR TITLE
matcher: add invert matching in metadata matcher

### PR DESCRIPTION
**Pull request by yangminzhu**
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

Commit Message: add invert matching in metadata matcher
Additional Description:
Risk Level: Low
Testing: Unit Test
Docs Changes: N/A
Release Notes: Updated version history
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/common/matchers.cc [hunk 1]
```diff
@@ -118,7 +118,7 @@ PathMatcher::createSafeRegex(const envoy::type::matcher::v3::RegexMatcher& regex
 
 bool MetadataMatcher::match(const envoy::config::core::v3::Metadata& metadata) const {
   const auto& value = Envoy::Config::Metadata::metadataValue(&metadata, matcher_.filter(), path_);
-  return value_matcher_ && value_matcher_->match(value);
+  return value_matcher_ && (value_matcher_->match(value) ^ matcher_.invert());
```
Comment (soulxu) [inline source/common/common/matchers.cc:121]: looks like `value_matcher_->match(value) ^ matcher_.invert()` is enough here.  `value_matcher_` should be always have value. anyway, it is nit.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM modulo @soulxu comment, thanks.

/wait

File: source/common/common/matchers.cc (hunk 1 above)
Comment (yangminzhu) [reply source/common/common/matchers.cc]: done.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Review by soulxu — APPROVED** [review]
LGTM, thanks for the update!

**Pull request merged**